src/main/scala/com/fulcrumgenomics/vcf/api/VcfHeader.scala

```diff
@@ -0,0 +1,175 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.vcf.api
+
+import com.fulcrumgenomics.FgBioDef._
+import enumeratum.EnumEntry
+import htsjdk.samtools.{SAMSequenceDictionary, SAMSequenceRecord}
+
+import scala.collection.immutable
+
+/** Trait used to represent how many instances of an attribute are to be expected in a VCFs INFO or genotype fields.*/
+sealed trait VcfCount { }
+
+object VcfCount {
+  /** VcfCount that indicates there should be exactly one value per alternate allele on the Variant. */
+  case object OnePerAltAllele  extends VcfCount
+  /** VcfCount that indicates there should be exactly one value per allele, including reference, on the Variant. */
+  case object OnePerAllele     extends VcfCount
+  /** VcfCount that indicates there should be exactly one value per possible genotype.  E.g. for a variant with a
+    * ref allele and one alt allele, and a diploid sample, there would be three values. */
+  case object OnePerGenotype   extends VcfCount
+  /** The attribute may have any number of values. */
+  case object Unknown          extends VcfCount
+  /** The attribute must have a fixed number of values, indicated by `count`. */
+  case class Fixed(count: Int) extends VcfCount
+}
+
+/** Trait for the set of acceptable atomic types that can appear in VCF's INFO and genotype fields. */
+sealed trait VcfFieldType extends EnumEntry {
+  /** Must parse a String into a single value of the type given. */
+  def parse(s: String): Any
+}
+
+object VcfFieldType extends FgBioEnum[VcfFieldType] {
+  /** Integer field type, represented internally as [[Int]]. */
+  case object Integer   extends VcfFieldType { override def parse(s: String): Any = s.toInt }
+  /** Floating point number field type, represented internally as [[Float]]. */
+  case object Float     extends VcfFieldType { override def parse(s: String): Any = s.toFloat }
+  /** String field type, represented internally as [[String]]. */
+  case object String    extends VcfFieldType { override def parse(s: String): Any = s }
+  /** Character field type, represented internally as [[Char]]. */
+  case object Character extends VcfFieldType { override def parse(s: String): Any = s.charAt(0) }
+  /** Character field type, represented internally as [[Char]]. */
+  case object Flag      extends VcfFieldType { override def parse(s: String): Any = "." }
+
+  override def values: immutable.IndexedSeq[VcfFieldType] = findValues
+}
+
+/** Trait representing an entry/line in a VCF header. */
+sealed trait VcfHeaderEntry {}
+
+// TODO add subclasses for ALT and PEDIGREE lines
```

Comment:
I think I could probably handle ALT.  But SAMPLE and PEDIGREE lines are so under-specified and I don't have experience with them, I'm not sure what those should look like.